File: src/lib/ai-insights.ts
import { fetchFromBackend } from "./api";

/**
 * Insights/predictions are now computed server-side (backend/services/ai_insights.py)
 * from the same real, RLS-scoped repository data the dashboards use — see the
 * fix plan's Phase 1. The role used to select a template is read from the
 * verified JWT on the backend, never from anything sent by this client.
 */

export interface InsightEvidence {
  label: string;
  detail: string;
  weight: number; // 0-100 contribution to the fused score
}

export interface RiskCase {
  id: string;
  subject: string;
  exam: string;
  level: "Low" | "Medium" | "High";
  score: number;
  evidence: InsightEvidence[];
}

export interface Insight {
  headline: string;
  body: string;
  action: { label: string; to: string } | null;
  /** Integrity: fused per-case risk instead of separate alerts. */
  cases?: RiskCase[];
  /** Staff warnings strip — omitted for students. */
  warnings?: { id: string; text: string; tone: "amber" | "rose" }[];
}

export interface Prediction {
  title: string;
  direction: "rising" | "stable" | "falling";
  summary: string;
  rows: {
    label: string;
    value: string;
    tone: "mint" | "amber" | "rose" | "iris";
  }[];
  action: { label: string; to: string } | null;
}

export function getInsight(): Promise<Insight | null> {
  return fetchFromBackend<Insight | null>("/api/insights");
}

export function getPrediction(): Promise<Prediction | null> {
  return fetchFromBackend<Prediction | null>("/api/predictions");
}

File: backend/schemas/ai_insights.py
from pydantic import BaseModel
from typing import List, Literal, Optional


class InsightEvidence(BaseModel):
    label: str
    detail: str
    weight: int


class RiskCase(BaseModel):
    id: str
    subject: str
    exam: str
    level: Literal["Low", "Medium", "High"]
    score: int
    evidence: List[InsightEvidence]


class InsightWarning(BaseModel):
    id: str
    text: str
    tone: Literal["amber", "rose"]


class InsightAction(BaseModel):
    label: str
    to: str


class Insight(BaseModel):
    headline: str
    body: str
    action: Optional[InsightAction] = None
    cases: Optional[List[RiskCase]] = None
    warnings: Optional[List[InsightWarning]] = None


class PredictionRow(BaseModel):
    label: str
    value: str
    tone: Literal["mint", "amber", "rose", "iris"]


class PredictionAction(BaseModel):
    label: str
    to: str


class Prediction(BaseModel):
    title: str
    direction: Literal["rising", "stable", "falling"]
    summary: str
    rows: List[PredictionRow]
    action: Optional[PredictionAction] = None


class RecommendationEvidence(BaseModel):
    label: str
    detail: str


class RecommendationBasedOn(BaseModel):
    source: str
    evidence: List[RecommendationEvidence]


class RecommendationAction(BaseModel):
    label: str
    to: Optional[str] = None
    confirmTitle: str
    confirmBody: str
    confirmLabel: str
    exports: Optional[bool] = None


class Recommendation(BaseModel):
    id: str
    kind: Literal["action", "guidance"]
    text: str
    basedOn: RecommendationBasedOn
    action: Optional[RecommendationAction] = None


class RecommendationSet(BaseModel):
    insightId: str
    items: List[Recommendation]


class ChatRequest(BaseModel):
    question: str


class ChatResponse(BaseModel):
    text: str
    blocked: bool
